src/ZEO/asyncio/client.py

```diff
@@ -132,60 +144,56 @@ def pop_futures(self):
         self.futures.clear()
         return futures
 
-    def protocol_factory(self):
-        return self
-
     def connect(self):
         if isinstance(self.addr, tuple):
             host, port = self.addr
-            cr = self.loop.create_connection(
+            cr = lambda: self.loop.create_connection(  # noqa: E731
                 self.protocol_factory, host or '127.0.0.1', port,
                 ssl=self.ssl, server_hostname=self.ssl_server_hostname)
         else:
-            cr = self.loop.create_unix_connection(
+            cr = lambda: self.loop.create_unix_connection(  # noqa: E731
                 self.protocol_factory, self.addr, ssl=self.ssl)
 
-        # an ``asyncio.Future``
-        self._connecting = cr = asyncio.ensure_future(cr, loop=self.loop)
-
-        @cr.add_done_callback
-        def done_connecting(future):
-            if future.cancelled():
-                logger.info("Connection to %r cancelled", self.addr)
-            elif future.exception() is not None:
-                logger.info("Connection to %r failed, %s",
-                            self.addr, future.exception())
-            else:
-                return
-
-            # keep trying
-            if not self.closed:
+        @coroutine
+        def connect():
+            while not self.closed:
+                try:
+                    yield cr()
```

Comment:
This should be `yield from` to be compatible with the typical "generator as coroutine" paradigm (same at many places).

Reply:
So you are pointing out to the core aspect of our CoroutineExecutor. Indeed `await` corresponds to `yield from`. async/await are actually plain generators with `yield from` wrapped into another syntax. But Python2 does not have support for `yield from`. But since we are already doing our own CoroutineExecutor, we can handle `yield core` with the same semantic as if it was `yield from coro`. This is handled by CoroutineExecutor here:

https://github.com/zopefoundation/ZEO/blob/0d0e16ab4490adf6aa3c8549b634b89baee83f4e/src/ZEO/asyncio/futures.py#L303-L306

in a sense we are implementing support for `yield from` ourselves.

This gives ability to follow the same style async/await is using with just a bit different syntax.